pkg/lightning/backend/remote/chunk_sender.go

```diff
@@ -0,0 +1,392 @@
+	ctx, cancel := context.WithCancel(ctx)
+	c := &chunkSender{
+		id: id,
+		wg: sync.WaitGroup{},
+
+		e:           engine,
+		httpClient:  engine.httpClient,
+		chunksCache: chunksCache,
+
+		state:       atomic.Pointer[chunksState]{},
+		nextChunkID: atomic.Uint64{},
+
+		loopDoneChan: make(chan struct{}),
+		taskChan:     make(chan *chunkTask),
+		cancel:       cancel,
+	}
+	c.state.Store(&chunksState{})
+
+	c.wg.Add(1)
+	go c.chunkSenderLoop(ctx)
+
+	engine.logger.Info("chunk sender started", zap.Uint64("sender", c.id))
+	return c
+}
+
+func (c *chunkSender) putChunk(ctx context.Context, data []byte) (uint64, error) {
+	task := newChunkTask(slices.Clone(data), false)
+
+	select {
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	case c.taskChan <- task:
+	case <-c.loopDoneChan:
+		return 0, c.err
+	}
+
+	select {
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	case newChunkID := <-task.resp:
+		return newChunkID, nil
+	case <-c.loopDoneChan:
+		return 0, c.err
+	}
+}
+
+func (c *chunkSender) flush(ctx context.Context) error {
+	if c.getFlushedChunkID() == c.getLastChunkID() {
+		return nil
+	}
+
+	task := newChunkTask(nil, true)
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case c.taskChan <- task:
+	case <-c.loopDoneChan:
+		return c.err
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-task.resp:
+	case <-c.loopDoneChan:
+		return c.err
+	}
+
+	return nil
+}
+
+func (c *chunkSender) getNextChunkID() uint64 {
+	return c.nextChunkID.Add(1)
+}
+
+func (c *chunkSender) getLastChunkID() uint64 {
+	return c.nextChunkID.Load()
+}
+
+func (c *chunkSender) getFlushedChunkID() uint64 {
+	return c.state.Load().FlushedChunkID
+}
+
+func (c *chunkSender) chunkSenderLoop(ctx context.Context) {
+	defer func() {
+		close(c.loopDoneChan)
+		err := c.chunksCache.close()
+		if err != nil {
+			c.e.logger.Warn("failed to close chunk cache", zap.Error(err))
+		}
+		c.wg.Done()
+	}()
+
+	ticker := time.NewTicker(updateFlushedChunkDuration)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case task := <-c.taskChan:
+			if !task.flush {
+				newChunkID := c.getNextChunkID()
+				task.done(newChunkID)
+
+				chunk := &chunk{
+					id:   newChunkID,
+					data: task.data,
+				}
+				c.err = c.putChunkToRemote(ctx, chunk)
+				if c.err != nil {
+					return
+				}
+			} else {
+				c.err = c.sendFlushToRemote(ctx)
+				if c.err != nil {
+					return
+				}
+				task.done(0)
+			}
+			// Reset the ticker when a new task is received. Avoid sending empty chunks too frequently.
+			ticker.Reset(updateFlushedChunkDuration)
+		case <-ticker.C:
+			// Periodically send empty chunks to update the flushed chunkID.
+			if c.getFlushedChunkID() == c.getLastChunkID() {
+				continue
+			}
+			c.err = c.putEmptyChunk(ctx)
+			if c.err != nil {
+				return
+			}
+		}
+	}
+}
+
+func (c *chunkSender) putEmptyChunk(ctx context.Context) error {
```

Comment:
what's the purpose of empty chunk? Please add comments

Reply:
Added the comment, PTAL~